onnxruntime/core/providers/vitisai/vitisai_custom_op.cc

```diff
@@ -0,0 +1,165 @@
+// Copyright(C) Xilinx Inc.
+// Licensed under the MIT License
+
+#include <fstream>
+#include <iostream>
+#include <string>
+
+#if defined(_MSC_VER)
+#pragma warning(disable : 4244 4245)
+#elif __GNUC__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+#endif
+#if defined(_MSC_VER)
+#pragma warning(default : 4244 4245)
+#elif __GNUC__
+#pragma GCC diagnostic pop
+#endif
+
+#include <pyxir/common/xbuffer.hpp>
+#include <pyxir/frontend/onnx.hpp>
+
+#include "vitisai_custom_op.h"
+#include "core/graph/graph_viewer.h"
+#include "core/graph/model.h"
+#include "core/common/logging/logging.h"
+#include "core/session/onnxruntime_cxx_api.h"
+
+namespace onnxruntime {
+namespace vitisai_ep {
+
+static ONNX_NAMESPACE::ModelProto GetModelProtoFromFusedNode(const onnxruntime::Node* fused_node,
+                                                             const logging::Logger& logger) {
+  const auto* node_function = fused_node->GetFunctionBody();
+
+  ORT_ENFORCE(node_function != nullptr, "Could not extract function body for node: ",
+              fused_node->Name());
+
+  const Graph& node_subgraph = node_function->Body();
+  onnxruntime::Model model{node_subgraph.Name(), true, ModelMetaData{}, PathString{},
+                           IOnnxRuntimeOpSchemaRegistryList{}, node_subgraph.DomainToVersionMap(),
+                           std::vector<ONNX_NAMESPACE::FunctionProto>(), logger};
+
+  ONNX_NAMESPACE::ModelProto model_proto = model.ToProto();
+  model_proto.set_ir_version(ONNX_NAMESPACE::Version::IR_VERSION);
+
+  *(model_proto.mutable_graph()) = node_subgraph.ToGraphProto();
+
+  return model_proto;
+}
+
+VitisAICustomOp::VitisAICustomOp(const ComputeContext* context,
+                         const onnxruntime::Node* fused_node,
+                         const logging::Logger* logger) 
+{
+  SetLogger(logger);
+
+  allocate_func_ = context->allocate_func;
+  release_func_ = context->release_func;
+  allocator_ = context->allocator_handle;
+  name_ = context->node_name;
+
+  ONNX_NAMESPACE::ModelProto model_proto = GetModelProtoFromFusedNode(fused_node, *GetLogger());
+  model_proto_ = model_proto;
+
```

Comment:
Probably gets optimized away in a release build but an unnecessary copy otherwise.